File: src/common/container/Flight.cpp
/***************************************************************************
 *   Copyright (C) 2004 by Alex Graf                                       *
 *   grafal@sourceforge.net                                                         *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include "Flight.h"
 
Flight::Flight()
{
	m_id = -1;
	m_number = -1;
	m_duration = -1;
	m_comment = "";
	m_distance = 0;
}

int Flight::id()
{
	return m_id;
}

void Flight::setId(int id)
{
	m_id = id;
}

Pilot& Flight::pilot()
{
	return m_pilot;
}

void Flight::setPilot(Pilot &pilot)
{
	m_pilot = pilot;
}

int Flight::number()
{
	return m_number;
}

void Flight::setNumber(int nr)
{
	m_number = nr;
}

const QDate& Flight::date()
{
	return m_date;
}

void Flight::setDate(const QDate &date)
{
	m_date = date;
}

const QTime& Flight::time()
{
	return m_time;
}

void Flight::setTime(const QTime &time)
{
	m_time = time;
}

WayPoint& Flight::startPt()
{
	return m_startPt;
}

void Flight::setStartPt(WayPoint &wp)
{
	m_startPt = wp;
}

WayPoint& Flight::landPt()
{
	return m_landPt;
}

void Flight::setLandPt(WayPoint &wp)
{
	m_landPt = wp;
}

int Flight::duration()
{
	return m_duration;
}

void Flight::setDuration(int sec)
{
	m_duration = sec;
}

void Flight::setDuration(const QTime &time)
{
	m_duration = time.hour() * 3600 + time.minute() * 60 + time.second();
}

Glider& Flight::glider()
{
	return m_glider;
}

void Flight::setGlider(Glider &glider)
{
	m_glider = glider;
}

const QByteArray& Flight::igcData() const
{
	return m_igcData;
}

void Flight::setIgcData(const QByteArray &data)
{
	m_igcData = data;
}

const QString& Flight::comment()
{
	return m_comment;
}

void Flight::setComment(const QString &comm)
{
	m_comment = comm;
}

uint Flight::distance()
{
	return m_distance;
}
 
void Flight::setDistance(uint dist)
{
	m_distance = dist;
}

Flight& Flight::operator=(const Flight &flight)
{
	m_id = flight.m_id;
	m_number = flight.m_number;
	m_pilot = flight.m_pilot;
	m_date = flight.m_date;
	m_time = flight.m_time;
	m_startPt = flight.m_startPt;
	m_landPt = flight.m_landPt;
	m_duration = flight.m_duration;
	m_glider = flight.m_glider;
	m_igcData = flight.m_igcData;
	m_comment = flight.m_comment;
	m_distance = flight.m_distance;
	
	return *this;
}

File: src/res/flight/js/flight.js
/***************************************************************************
 *   Copyright (C) 2012 by Alex Graf                                       *
 *   grafal@sourceforge.net                                                *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 *                                                                         *
 *   It is prohibited to serve or run this code over network p.e. as web   *
 *   service in combination with closed source.                            *
 ***************************************************************************/

function Flight(map)
{
	var flight = this;

	this.map = map;

	this.glider = new google.maps.Marker({
		map: map,
		draggable: true,
		raiseOnDrag: false,
		zIndex: 15,
		icon: 'http://chart.apis.google.com/chart?cht=mm&chs=20x32&chco=FFFFFF,FFFFFF,000000&ext=.png'
	});

	this.track = new google.maps.Polyline({
		strokeColor: '#FFBB00',
		strokeOpacity: 1.0,
		strokeWeight: 3,
		map: map,
		zIndex: 1
	});

	this.stpos = null;
	this.trackPts = null;
	this.sogList = null;
	this.varioList = null;
	this.altList = null;
	this.minAlt = 0;
	this.maxAlt = 0;
	this.timeList = null;

	google.maps.event.addListener(this.glider, 'dragstart', function(event) {fl_dragstart(flight);});
	google.maps.event.addListener(this.glider, 'drag', function(event) {fl_drag(flight);});
}

Flight.prototype.getMap = function()
{
	return this.map;
}

Flight.prototype.setTrackPts = function(trackPts)
{
	this.trackPts = trackPts;
	this.track.setPath(trackPts);
}

Flight.prototype.getTrackPts = function()
{
	return this.trackPts;
}

Flight.prototype.setSogList = function(sogList)
{
	this.sogList = sogList;
}

Flight.prototype.getSogAt = function(index)
{
	var sog = null;

	if((this.sogList != null) && (index < this.sogList.length))
	{
		sog = this.sogList[index];
	}

	return sog;
}

Flight.prototype.setVarioList = function(varioList)
{
	this.varioList = varioList;
}

Flight.prototype.getVarioAt = function(index)
{
	var vario = null;

	if((this.varioList != null) && (index < this.varioList.length))
	{
		vario = this.varioList[index];
	}

	return vario;
}

Flight.prototype.setAltList= function(altList, minAlt, maxAlt)
{
	this.altList = altList;
	this.minAlt = minAlt;
	this.maxAlt = maxAlt;
}

Flight.prototype.getAltList= function()
{
	return this.altList;
}

Flight.prototype.getAltAt = function(index)
{
	var alt = null;

	if((this.altList != null) && (index < this.altList.length))
	{
		alt = this.altList[index];
	}

	return alt;
}

Flight.prototype.getMinAlt = function()
{
	return this.minAlt;
}

Flight.prototype.getMaxAlt = function()
{
	return this.maxAlt;
}

Flight.prototype.setTimeList = function(timeList)
{
	this.timeList = timeList;
}

Flight.prototype.getTimeList = function()
{
	return this.timeList;
}

Flight.prototype.getTimeAt = function(index)
{
	var time = null;

	if((this.timeList != null) && (index < this.timeList.length))
	{
		time = this.timeList[index];
	}

	return time;
}

Flight.prototype.moveGliderTo = function(index)
{
	var time = null;

	if((this.trackPts != null) && (index < this.trackPts.length))
	{
		this.glider.setPosition(this.trackPts[index]);
	}
}

/*
	This is an ugly hack, to restore position while drag. Because Qt 4.6 won't display
	markers which are not draggable. In a later version, this should be fixed through
	setting markers draggable=this.editable.
*/
Flight.prototype.storePos = function()
{
	this.stpos = new google.maps.LatLng(this.glider.getPosition().lat(), this.glider.getPosition().lng());
};

Flight.prototype.restorePos = function()
{
	if(this.storePos != null)
	{
		this.glider.setPosition(this.stpos);
	}
}

function fl_dragstart(flight)
{
	flight.storePos();
}

function fl_drag(flight)
{
	flight.restorePos();
}
